pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,282 @@
+	return &NodeLabeller{
+		kvmController:    kvmController,
+		kubevirtInformer: kubevirtInformer,
+		nodeInformer:     nodeInformer,
+		clientset:        clientset,
+		Queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:           host,
+		namespace:      namespace,
+		logger:         log.DefaultLogger(),
+		clusterConfig:  clusterConfig,
+		virshCommandFN: runVirshCommand,
+		capScanner:     NewCapScanner(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	n.cpuInfo.Lock()
+	err := n.loadCPUInfo()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load cpu info")
+	}
+	n.cpuInfo.Unlock()
+
+	n.hostCapabilities.Lock()
+	err = n.loadHostCapabilities()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load host capabilities")
+	}
+	n.hostCapabilities.Unlock()
+
+	n.hypervFeatures.Lock()
+	n.hypervFeatures.items = n.capScanner.getLabels()
+	n.hypervFeatures.Unlock()
+
+	n.kubevirtInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) {
+			n.Queue.Add(n.host)
+		},
+		DeleteFunc: func(obj interface{}) {
+			_, ok := obj.(*v1.ConfigMap)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
+					return
+				}
+				_, ok = tombstone.Obj.(*v1.ConfigMap)
+				if !ok {
+					log.Log.Reason(fmt.Errorf("tombstone contained object that is not a config map %#v", obj)).Error("Failed to process delete notification")
+					return
+				}
+			}
+
+			n.Queue.Add(n.host)
+		},
+		UpdateFunc: func(_, _ interface{}) {
+			n.Queue.Add(n.host)
+		},
+	})
+
+	go n.kubevirtInformer.Run(stop)
+
+	n.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, _ interface{}) {
+			n.Queue.Add(n.host)
+		},
+	})
+
+	go n.nodeInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.nodeInformer.HasSynced, n.kubevirtInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+
+	if !n.kvmController.NodeHasDevice(kvmPath) {
+		n.logger.Errorf("Node-labeller cannot work without KVM device.")
+		return true
+	}
+
+	err := n.run()
+
+	if err != nil {
+		n.logger.Errorf("node-labeller sync error encountered: %v", err)
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
```

Comment:
It is not needed to `make` an empty map and slice here, because they will be overwritten in the following line. The compiler may be able to optimize it out, but it is useless code.